### PR TITLE
ACROSS API: Add underlying API code

### PR DESCRIPTION
**Pull request by jak574**
# ACROSS API 

## Intro

The ACROSS API is a RESTful API designed to access up-to-date state information on NASA and other missions involved in Time Domain and Multi-Messenger Astrophysics (TDAMM) science. The aim of the API is to enable easier coordination of observations with NASA's fleet of missions initially, with aims to expand beyond this scope in later years.

The ACROSS API is written in Python and utilizes the [FastAPI](https://fastapi.tiangolo.com) and [Pydantic](https://docs.pydantic.dev/latest/) libraries to produce an API that is OpenAPI 3.1 compliant. This includes auto documentation of the API interface. It is designed to run on AWS as a Lambda function, and be integrated into the NASA GCN website, under the gcn.nasa.gov/labs route.  

## This Pull Request

This pull request contains the base-level code that enables the ACROSS API. This can be divided into four main areas:
* The FastAPI / Pydantic based core of the the ACROSS API on which the API is built.
* The method for caching API results using DynamoDB tables and if necessary S3 bucket for cached results > 400kBytes.
* The method for authenticating with username / api_key for tasks that require this, using a User table stored in a DynamoDB table, and a decorator around the `get`/`post`/`put`/`del` method for the API class that specifies the required authentication.
* Two sample API endpoints that perform the following tasks:
  - `hello` - replies with a welcome message, with your name as a optional supplied argument
  - `Resolve` - Given a valid astronomical source name, this API call resolves it into coordinates using either Simbad or ANTARES for ZTF source names.

## Aim

The aim of this pull request is to start the process of integration of the ACROSS API with GCN. To this end, this initial pull contains an extremely bare-bones version of the current ACROSS API development code, with the expectation that future pull requests will incrementally add in features. The first of these features will pertain to the development of the BurstCube "Target of Opportunity" form.

## Outstanding issues

Here I list any issues that may need to be resolved. 
* Currently if job caching needs to cache a result > 400kBytes, it needs to store this in an S3 bucket, as this is too large for a DynamoDB table. The creation and access of this bucket needs to be corrected to work correctly with the GCN AWS setup.

### CODE REVIEW

**Review by lpsinger — COMMENTED** [review]
At a first look, the code is well documented. This is a very large PR, so it will take some time for me to review it.

I notice that you have added some pragma comments related to type checking. What type checker would you like us to add to the CI/CD pipeline?

One very high level comment is that we are going to have to replace the authentication code with OIDC access tokens.

File: python/across_api/_version.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# file generated by setuptools_scm
+# don't change, don't track in version control
```
Comment (lpsinger) [inline python/across_api/_version.py:2]: We aren't currently tagging releases, so I'm not sure that setuptools_scm will be very functional. Can we remove this?

File: python/requirements.txt [hunk 2]
```diff
@@ -1,3 +1,45 @@
...
+s3transfer==0.8.2
+sgp4==2.23
+shapely==2.0.2
+simplejson==3.19.2
+six==1.16.0
+sniffio==1.3.0
+starlette==0.27.0
+typing_extensions==4.9.0
+urllib3==2.0.7
+xmltodict==0.13.0
```
Comment (lpsinger) [inline python/requirements.txt:45]: Ah, we should set up Dependabot for Python, shouldn't we?

**Comment by jak574** [comment]
Regarding type checking, I'm using mypy, with the [Pydantic mypy plugin](https://docs.pydantic.dev/latest/integrations/mypy/) installed.

File: python/across_api/_version.py (hunk 1 above)
Comment (jak574) [reply python/across_api/_version.py:2]: Yes, I'll remove it.

Comment (jak574) [reply python/across_api/_version.py]: Check in  258bfae fixes this.

Comment (jak574) [reply python/across_api/_version.py]: [52c5fd2](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1742/commits/52c5fd2704cc0a05e8780c28cdcc0a72137f5bb6)

**Comment by lpsinger** [comment]
> Regarding type checking, I'm using mypy, with the [Pydantic mypy plugin](https://docs.pydantic.dev/latest/integrations/mypy/) installed.

Do you have the bandwidth right now to do some shopping for a GitHub action for that?

**Comment by jak574** [comment]
Yeah, I'll look into the mypy workflow.

**Comment by jak574** [comment]
OK I added a basic mypy check to the python workflow. I'll look into how to add the Pydantic module to the check.

File: python/across_api/across/resolve.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    dra = None
+    for line in lines:
+        x = line.split("|")
+        try:
+            name = x[0]
+            numbers = x[1].split(" ")
+            dra = float(numbers[1])
+            ddec = float(numbers[2].strip())
+        except (ValueError, IndexError):
+            pass
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:77]: Replace with [stropy.coordinates.SkyCoord.from_name](https://docs.astropy.org/en/stable/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.from_name).

File: python/across_api/across/api.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+@app.get("/")
+async def hello(name: YourNameDep) -> HelloSchema:
+    """
+    This function returns a JSON response with a greeting message and an optional name parameter.
+    If the name parameter is provided, the greeting message will include the name.
+    """
+    return Hello(name=name).schema
+
+
+@app.get("/ACROSS/Resolve")
```
Comment (lpsinger) [inline python/across_api/across/api.py:46]: It's a bit unusual to have uppercase characters in URLs.

File: python/across_api/across/resolve.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            return False
+
+        """Do a name search"""
+        # Check against the ANTARES broker
+        if "ZTF" in self.name:
+            ra, dec = antares_radec(self.name)
+            if ra is not None:
+                self.ra, self.dec = ra, dec
+                self.resolver = "ANTARES"
+                return True
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:158]: Consider using https://pypi.org/project/antares-client/.

File: .github/workflows/pull_request.yml [hunk 6]
```diff
@@ -26,5 +26,9 @@ jobs:
       - name: Pytest
         run: pytest python
 
+      - name: Mypy action
+        run: |
+          mypy  $(git ls-files '*.py')
+
```
Comment (lpsinger) [inline .github/workflows/pull_request.yml:32]: Please move this to a separate PR. Also, we need to add mypy to both the GitHub actions _and_ the pre-commit hook.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: python/across_api/across/resolve.py (hunk 5 above)
Comment (jak574) [reply python/across_api/across/resolve.py:158]: There's a reason why I don't. It installs a bunch of huge dependencies like pandas that I otherwise don't need. Given the limitations on size with AWS Lambda, and the fact that I only use this one tiny part of the ANTARES API, this is way leaner.

File: .github/workflows/pull_request.yml (hunk 6 above)
Comment (jak574) [reply .github/workflows/pull_request.yml:32]: OK.

File: python/across_api/across/api.py (hunk 4 above)
Comment (jak574) [reply python/across_api/across/api.py]: Fixed in 893f9c1.

File: python/across_api/across/resolve.py (hunk 3 above)
Comment (jak574) [reply python/across_api/across/resolve.py:77]: I actually tried to do this before but it crashes for reasons I don't understand.

Comment (jak574) [reply python/across_api/across/resolve.py:77]: The error:

```  File "...across_api/across/resolve.py", line 133, in get
    skycoord = SkyCoord.from_name(self.name)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "...vendor/astropy/coordinates/sky_coordinate.py", line 2205, in from_name
    icrs_coord = get_icrs_coordinates(name, parse, cache=cache)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "...vendor/astropy/coordinates/name_resolve.py", line 207, in get_icrs_coordinates
    sc = SkyCoord(ra=ra, dec=dec, unit=(u.degree, u.degree), frame="icrs")
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "...vendor/astropy/coordinates/sky_coordinate.py", line 333, in __init__
    frame_cls, frame_kwargs = _get_frame_without_data(args, kwargs)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "...vendor/astropy/coordinates/sky_coordinate_parsers.py", line 143, in _get_frame_without_data
    frame_cls = _get_frame_class(frame)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "...vendor/astropy/coordinates/sky_coordinate_parsers.py", line 47, in _get_frame_class
    raise ValueError(
ValueError: Coordinate frame name "icrs" is not a known coordinate frame ([])
```

Comment (jak574) [reply python/across_api/across/resolve.py]: Bug resolved, so 6cfaa75 fixes this.

File: python/across_api/across/resolve.py (hunk 5 above)
Comment (lpsinger) [reply python/across_api/across/resolve.py:127]: What dependencies are there other than pandas? We are quite likely to acquire a dependency on that anyway.

Comment (jak574) [reply python/across_api/across/resolve.py:127]: It adds pandas (~60Mb), marshmallow, confluent-kafka and a few other small things. One issue I ran into is that its version of confluent-kafka is required to be 1.70, which conflicts with pygcn which requires >2.2.0. This isn't important for now as I don't use pygcn in the API (yet), however for whatever reason, I cannot get this old version of confluent-kafka to build on either of my dev machines, so it completely breaks my dev environment.

Comment (lpsinger) [reply python/across_api/across/resolve.py:127]: The [quota for Lambda function size](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html#compute-and-storage) is 75 GB and can be adjusted up to terabytes upon request. Avoiding a 60 MB dependency is really premature optimization.

Would you please submit a PR upstream to antares_client to unpin their version of confluent-kafka?

Comment (jak574) [reply python/across_api/across/resolve.py:127]: Lambda zip-file deployment size is max of ~250Mb (uncompressed), and I hit this easily when doing testing, hence why I made this simple code to call the ANTARES REST API. I'm fine making this change, as it's easy to do (my original code used antares-client before I hit these quota issues). I'll keep the code in my back pocket if we have quota issues.

Comment (jak574) [reply python/across_api/across/resolve.py:114]: OK, pushed the change in caccb8b.

File: python/requirements.txt [hunk 7]
```diff
@@ -1,3 +1,47 @@
...
+ecdsa==0.18.0
+email-validator==2.1.0.post1
+fastapi==0.105.0
+idna==3.6
+jmespath==1.0.1
+mangum==0.17.0
+numpy==1.26.2
+packaging==23.2
+pipdeptree==2.13.1
+psycopg2-binary==2.9.9
```
Comment (lpsinger) [inline python/requirements.txt:24]: I doubt we use either of these packages. Where does this list of dependencies come from?

Comment (jak574) [reply python/requirements.txt:24]: pipdeptree one of yours from the original requirements.txt. psycopg2-binary is because stuff not part of this pull request (but coming in later one) uses a PostgreSQL database. I did not pull out all the unused requirements for this pull request, I will admit.

Comment (jak574) [reply python/requirements.txt]: Anyway... I pushed bc073f9 which updates requirements.txt to include only modules needed for this pull request.

File: python/requirements.txt [hunk 8]
```diff
@@ -1,3 +1,48 @@
...
+astropy==6.0.0
+astropy-iers-data
+boto3==1.34.2
+botocore==1.34.2
+bson==0.5.10
+certifi==2023.11.17
+cffi==1.16.0
+charset-normalizer==3.3.2
+click==8.1.7
+confluent-kafka==1.7.0
```
Comment (lpsinger) [inline python/requirements.txt:14]: I don't see any use of confluent-kafka here. Why is it listed as a dependency?

Comment (jak574) [reply python/requirements.txt:14]: It's a dependency of antares-client.

Comment (jak574) [reply python/requirements.txt:14]: ...and I see the same issue I had is causing the python github workflow to fail.

Comment (lpsinger) [reply python/requirements.txt:14]: Right, so the CI/CD pipeline is failing because there aren't wheels on PyPI for that version of confluent-kafka-python. I suggest opening a PR with antares_client to unpin their dependency so that downstream projects (like us) can use more recent versions.

But what I'm trying to get at, is how are you generating this list of dependencies? Is this a lock file representing everything in your Python environment, captured with `pip freeze`?

Comment (jak574) [reply python/requirements.txt:14]: I'm creating a new Python virtualenv, installing only the modules needed for the API to work, and then dumping them to `requirements.txt` with `pip freeze`, yes. 

I submitted an issue to the ANTARES client GitLab page about the version of confluent-kafka they're using.

Comment (lpsinger) [reply python/requirements.txt:14]: > I'm creating a new Python virtualenv, installing only the modules needed for the API to work, and then dumping them to `requirements.txt` with `pip freeze`, yes.

Alright. So up until now I have been lazy about this because we had very few dependencies, but I would now like to pick a Python dependency management tool that provides functionality comparable to NPM's package.json (specifying constraints on compatible versions of dependencies) and package-lock.json (specifying concrete resolved versions of dependencies). So we need to do a miniature trade study for that and pick one, probably in a new issue and PR.

> I submitted an issue to the ANTARES client GitLab page about the version of confluent-kafka they're using.

Thank you. Would you please add a link to it here?

Comment (jak574) [reply python/requirements.txt:14]: https://gitlab.com/nsf-noirlab/csdc/antares/client/-/issues/8

Comment (lpsinger) [reply python/requirements.txt:14]: https://gitlab.com/nsf-noirlab/csdc/antares/client/-/merge_requests/32

Comment (jak574) [reply python/requirements.txt:14]: @lpsinger this comment probably sheds some light as to why they pin the dependency: https://gitlab.com/nsf-noirlab/csdc/antares/client/-/merge_requests/22#note_1532581942

Comment (lpsinger) [reply python/requirements.txt:14]: Pretty sure that is fixed upstream by now, as the most recent versions of confluent-kafka-python definitely do validate TLS certificates.

What problem is this currently causing right now? The Lambda doesn't have a dependency on gcn-kafka-python at the moment, does it?

**Comment by lpsinger** [comment]
> OK I added a basic mypy check to the python workflow. I'll look into how to add the Pydantic module to the check.

See #1759.

File: python/requirements.txt (hunk 8 above)
Comment (jak574) [reply python/requirements.txt:14]: Right now the problem is that antares-client won't build in the CI/CD workflow, because of the dependency on an old version of confluent-kafka module. This is because it compiles from source and needs librdkafka installed to build correctly.

**Comment by lpsinger** [comment]
As I said, the code quality here (code style, comments and docstrings, comprehensibility) is very high here, kudos! I am still going through this line by line and I'll have more minor comments of the kind I've already posted. However, I have some high-level structural comments and questions:

- We need to move the auth from API keys to OIDC access tokens. This will probably not be a huge lift given that you have had the foresight to implement the auth in a way that is cleanly orthogonal to the API calls themselves. (Side note: it may improve compatibility with other projects if we adopt https://scitokens.org for how we name the OIDC claims.)
- Much of the schema and API introduced here models submission and tracking of long-running jobs submitted via a REST API.
  - Are we sure that there is no prior art for that?
  - How long do we expect jobs to run and how resource-intensive are they? What is an example of the longest-running background job that we expect?
  - Do we have any ideas how we are going to queue and execute these jobs? If not, I would suggest AWS Lambda for execution because of its easy deployment and resource management, and Amazon SNS for queuing because it has built-in support in Architect and is integrated with Lambda.

**Comment by jak574** [comment]
> As I said, the code quality here (code style, comments and docstrings, comprehensibility) is very high here, kudos! I am still going through this line by line and I'll have more minor comments of the kind I've already posted. However, I have some high-level structural comments and questions:
> 
> * We need to move the auth from API keys to OIDC access tokens. This will probably not be a huge lift given that you have had the foresight to implement the auth in a way that is cleanly orthogonal to the API calls themselves. (Side note: it may improve compatibility with other projects if we adopt https://scitokens.org for how we name the OIDC claims.)

Agreed. I can just pull the auth code in there right now, as nothing in this pull request requires authentication, and we put it back in using OIDC in a later pull.

> * Much of the schema and API introduced here models submission and tracking of long-running jobs submitted via a REST API.

Some of this is hold overs from the Swift API, which is queue based. Again as with authentication the implementation of this here is through decorator, so can easily be swapped out for a different implementation. Mostly I see the need for two things:
1. Caching of results from large, common jobs that are CPU intensive or executed so regularly, that's it's best to not run over and over again (e.g. generating an ephemeris).
2. Logging of events, which right now this doesn't really do (in a previous incarnation using a regular SQL database, job caches doubled as logs), the DynamoDB version looses that as it uses job parameters as a partition key. 

>   * Are we sure that there is no prior art for that?

For job caching, I would assume there is prior art yes. We could for example use https://github.com/long2ice/fastapi-cache for caching requests at the API level. 

>   * How long do we expect jobs to run and how resource-intensive are they? What is an example of the longest-running background job that we expect?

Probably the longest jobs right now are generating ephemerides which takes a few seconds to run. 

>   * Do we have any ideas how we are going to queue and execute these jobs? If not, I would suggest AWS Lambda for execution because of its easy deployment and resource management, and Amazon SNS for queuing because it has built-in support in Architect and is integrated with Lambda.

OK, I'm unfamiliar with SNS so will have to look into it.

File: python/requirements.txt (hunk 8 above)
Comment (lpsinger) [reply python/requirements.txt:14]: Ahhh, I see.

Comment (lpsinger) [reply python/requirements.txt:14]: OK, in that case, let's use your original code for Antares. But would you please add a FIXME comment explaining the problem and including a link to the upstream issue?

**Comment by jak574** [comment]
> Or, we could set HTTP cache-control headers. Eligible responses will be cached by our content delivery network at edge locations close to users.

OK, so there's an issue with using caching like this. The job caching is used inside the API code, so for example, if you want to calculate visibility for a target, the API will call the code to generate the ephemeris (which is cached), and then uses that ephemeris data to calculate visibility. So if a user wants to calculate the visibility of 100 targets for a specific time period, the ephemeris is only generated once. 

As the API handles the caching at the decorator level, this means the internal API calls are cached, so repeated calls to `Ephem` class with the same arguments return a cached result. 

If we want to do this in a "not roll your own" way, then something like https://github.com/jaymon/caches might be the solution here.

> For anything under about 15 seconds, I think it's better to simply return the response to the client directly. Do we anticipate any workflows that will run much longer than that? If so, what are they?

For an individual API call I would not expect >15s, but I suppose it's possible. It depends on what limitations we put on queries. For example, you can imagine someone could put in a request to calculate the visibility of target for a year. This would involve generating a year's worth of ephemerides. With the Swift API I limit length of time allowed to stop this sort of thing. 

> For what it's worth, the maximum runtime for API responses from the web site is normally 30 seconds, which is the [maximum integration timeout for API Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/limits.html). Lambdas themselves _can_ run for a lot longer than that, up to 15 minutes.

I think as a general rule we want our API to be responsive, and not something that can take 15 mins to respond. I think we should therefore not allow these kinds of requests.

**Review by jak574 — COMMENTED** [review]
Turns out this does not work with FASTAPI.

**Comment by jak574** [comment]
@lpsinger I think I've addressed everything that we discussed at the review. Regarding caching, I need to think about this some more, so have just removed it completely for now.

File: python/across_api/across/user.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
+from secrets import token_urlsafe
```
Comment (lpsinger) [inline python/across_api/across/user.py:1]: I think that most of this file will not survive the transition to OIDC.

File: python/across_api/across/resolve.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        try:
+            skycoord = SkyCoord.from_name(self.name)
+            self.ra, self.dec = skycoord.ra.deg, skycoord.dec.deg
+            self.resolver = "CDS"
+            return True
+        except NameResolveError:
+            pass
+
+        # Send a warning if name couldn't be resolved
+        raise HTTPException(status_code=404, detail="Could not resolve name.")
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:132]: Hmm, 404 not found doesn't seem like the right status code here.

File: python/across_api/across/models.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        The API key associated with the user.
+    userlevel: int
+        The user level. Defaults to 1.
+    """
+
+    __tablename__ = "acrossapi_users"
+
+    username: str
+    api_key: str
+    userlevel: int = 1
```
Comment (lpsinger) [inline python/across_api/across/models.py:22]: Remove... won't apply to OIDC.

File: python/across_api/across/hello.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+from .schema import HelloGetSchema, HelloSchema
+
+
+class Hello(ACROSSAPIBase):
+    """Sample ACROSS API Class
+
+    Parameters
+    ----------
+    name : str
+        Your name, if you wish to give it
```
Comment (lpsinger) [inline python/across_api/across/hello.py:13]: Is it conventional that if we are annotating types that we _also_ need to supply the types in Numydoc-style docstrings? Seems like there should be a way to avoid repeating ourselves.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: python/across_api/base/schema.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+class BaseSchema(BaseModel):
+    """
+    Base class for schemas.
+
+    This class provides a base implementation for schemas and defines the `from_attributes` method.
+    Subclasses can inherit from this class and override the `from_attributes` method to define their own schema logic.
+    """
+
+    model_config = ConfigDict(from_attributes=True)
```
[the remark this answers is not in the document]
Comment (jak574) [reply python/across_api/base/schema.py:12]: I'm doing this to universally apply the ability to create Pydantic models from class attributes. 

https://docs.pydantic.dev/latest/concepts/models/#arbitrary-class-instances

This is useful because it avoids the need to covert objects into dicts when validating them against a Pydantic model. Generally it makes the code cleaner IMHO.

File: python/across_api/across/resolve.py (hunk 10 above)
Comment (jak574) [reply python/across_api/across/resolve.py:132]: Haha! I did agonize over this one. 404 is probably a stretch I'll admit, but it's not clear to which which status code to communicate that the request did not work because the name of the object you typed in was not recognized or "Not Found". Suggestions welcome.

File: python/across_api/across/models.py (hunk 11 above)
Comment (jak574) [reply python/across_api/across/models.py:22]: OK, I will just pull out all the auth code.

File: python/across_api/across/hello.py (hunk 12 above)
Comment (jak574) [reply python/across_api/across/hello.py:13]: The [numpy docstring style guide](https://numpydoc.readthedocs.io/en/latest/format.html) says to indicate type if it is specified, and in strongly typed code like this I would say we should always put in type. This does place the burden of making sure code and comments are consistent of course. However, I would suggest that the current usage is the correct one.

File: python/across_api/base/api.py [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+        ),
+    ] = None,
+) -> dict:
+    """
+    Helper function to convert begin and end to datetime objects.
+    """
+    return {"begin": convert_to_dt(begin), "end": convert_to_dt(end)}
+
+
+OptionalDateRangeDep = Annotated[dict, Depends(optional_daterange)]
```
[the remark this answers is not in the document]
Comment (jak574) [reply python/across_api/base/api.py]: Fixed in ba9c4f5.

File: python/across_api/across/user.py (hunk 9 above)
Comment (jak574) [reply python/across_api/across/user.py:1]: Removal of auth code will resolve this.

File: python/across_api/across/models.py (hunk 11 above)
Comment (jak574) [reply python/across_api/across/models.py]: Fixed by cdf0e9f

File: python/across_api/across/user.py (hunk 9 above)
Comment (jak574) [reply python/across_api/across/user.py]: Fixed by cdf0e9f

File: python/across_api/version.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
+__version__ = version = "1.0"
+__version_tuple__ = version_tuple = (1, 0)
+
+API_VERSION = f"{__version_tuple__[0]}.{__version_tuple__[1]}"
```
[the remark this answers is not in the document]
Comment (jak574) [reply python/across_api/version.py]: Removed. See aadc267.

File: python/across_api/across/resolve.py (hunk 10 above)
Comment (jak574) [reply python/across_api/across/resolve.py:132]: BTW it strikes me that 404 might be appropriate if we change the resolve endpoint to work like this:

`/across/resolve/Crab`

rather than:

`/across/resolve?name=Crab`

For the former I think a 404 is appropriate, and it's a small change.

Comment (lpsinger) [reply python/across_api/across/resolve.py:132]: How about just returning a JSON response that omits the ra and dec properties?

Comment (jak574) [reply python/across_api/across/resolve.py]: b927619 reverts the code to how I handle this in the Swift API, which does what you suggest.

File: python/across_api/across/hello.py (hunk 12 above)
Comment (jak574) [reply python/across_api/across/hello.py]: a8b1c4b

File: python/across_api/across/resolve.py [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        {"query": {"bool": {"filter": {"term": {"properties.ztf_object_id": ztf_id}}}}}
+    )
+
+    params = {
+        "sort": "-properties.newest_alert_observation_time",
+        "elasticsearch_query[locus_listing]": search_query,
+    }
+    r = requests.get(ANTARES_URL, params=params)
+
+    if r.status_code == 200:
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:39]:
```suggestion
    if r.ok:
```

File: python/across_api/across/resolve.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        antares_data = json.loads(r.text)
+        ra = antares_data["data"][0]["attributes"]["ra"]
+        dec = antares_data["data"][0]["attributes"]["dec"]
+        return ra, dec
+    else:
+        return None, None
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:45]: Perhaps we should log the error as well?

File: python/across_api/across/resolve.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        try:
+            skycoord = SkyCoord.from_name(self.name)
+            self.ra, self.dec = skycoord.ra.deg, skycoord.dec.deg
+            self.resolver = "CDS"
+            return True
+        except NameResolveError:
+            pass
+
+        # If no resolution occurred, report this to the user
+        self.resolver = "Could not resolve name."
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:124]: How about just returning no resolver, or None? I'm always hesitant to signal errors "in band" by using values that are allowed as normal return types --- in this case a string.

File: python/across_api/across/resolve.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        return False
+
+
+ACROSSAPIResolve = Resolve
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:128]: What is the purpose of this alias?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: python/across_api/across/resolve.py (hunk 17 above)
Comment (jak574) [reply python/across_api/across/resolve.py:45]: I don't really understand what you mean?

File: python/requirements.in [hunk 20]
```diff
@@ -0,0 +1,4 @@
+astropy==6.0.0
+fastapi==0.105.0
+mangum==0.17.0
+requests==2.31.0
```
[the remark this answers is not in the document]
Comment (jak574) [reply python/requirements.in:4]: Explain please.

File: python/across_api/across/resolve.py (hunk 18 above)
Comment (jak574) [reply python/across_api/across/resolve.py:124]: I'm even less of a fan of code failing silently. We either need to report an error code or add some sort of passing of errors/warnings. The latter did exist before I removed the Job stuff.

File: python/requirements.in (hunk 20 above)
[the remark this answers is not in the document]
Comment (lpsinger) [reply python/requirements.in:4]: For Python (JavaScript), the requirements.in (package.json) file defines direct dependencies of the project and constrains the range of compatible versions of those dependencies. The requirements.txt (package-lock.json) file, which is automatically generated from the former file, captures a particular dependency solution with concrete version numbers. Generally, the requirements.in file should express ranges of versions but not literal version numbers.

For example, we are probably going to require a version of Astropy that is API compatible with 6.0.0 because of my patches that went into that release. According to [SemVer](https://semver.org) rules, a compatible version would be greater than or equal to 6.0.0 but less than 7.0.0. So in requirements.in, we might express that by writing `astropy~=6.0.0` or (roughly equivalently) `astropy>=6.0.0,<7`.

[the remark this answers is not in the document]
Comment (jak574) [reply python/requirements.in:4]: FastAPI suggests that you pin the version (https://fastapi.tiangolo.com/deployment/versions/) for deployment.

File: python/across_api/across/resolve.py (hunk 18 above)
Comment (lpsinger) [reply python/across_api/across/resolve.py:124]: I think that returning None for ra, dec, and resolver is a pretty clear indication that nothing was found. Especially if we document that.

File: python/across_api/across/resolve.py (hunk 17 above)
Comment (jak574) [reply python/across_api/across/resolve.py:45]: OK, logging was something I wanted to address in a later pull request.

File: python/across_api/across/resolve.py (hunk 18 above)
Comment (jak574) [reply python/across_api/across/resolve.py]: 3c0a8ce

File: python/requirements.in (hunk 20 above)
[the remark this answers is not in the document]
Comment (lpsinger) [reply python/requirements.in:4]: Well, yes, and we do --- in requirements.txt. But don't pin it in requirements.in.

If and when you find that we are depending on a specific feature of FastAPI that was added in version X.Y.Z, then you can update requirements.in to say: `fastapi>=X.Y.Z`.

File: python/across_api/across/resolve.py (hunk 17 above)
Comment (lpsinger) [reply python/across_api/across/resolve.py:45]: Understandable, but unfortunately it's necessary here.

File: python/requirements.in (hunk 20 above)
[the remark this answers is not in the document]
Comment (jak574) [reply python/requirements.in:4]: How often do you envisage that we'll run pip-compile then?

[the remark this answers is not in the document]
Comment (lpsinger) [reply python/requirements.in:4]: We'll run pip-compile _by hand_ pretty infrequently --- only when we are adding dependencies.

However, every week Dependabot will scan our repository's dependencies and open PRs to update any of our dependencies that have new releases. We'll be able to review those PRs and either merge them (if the risk is low and any tests pass) or reject them (if the risk of updating is high or the tests fail).

[the remark this answers is not in the document]
Comment (lpsinger) [reply python/requirements.in:4]: You can see what those Dependabot PRs look like here: https://github.com/nasa-gcn/gcn.nasa.gov/pulls?q=is%3Apr+author%3Aapp%2Fdependabot

File: python/across_api/across/resolve.py (hunk 17 above)
Comment (jak574) [reply python/across_api/across/resolve.py]: c44c0ba

File: python/requirements.in (hunk 20 above)
[the remark this answers is not in the document]
Comment (jak574) [reply python/requirements.in]: OK I guess I misunderstood the interplay between requirements.in and Dependabot. I assume it'd be updating the values in that file, then would generate requirements.txt from it. 

So anyway, I unpinned in this commit c177e84.

File: python/across_api/across/resolve.py [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+        "sort": "-properties.newest_alert_observation_time",
+        "elasticsearch_query[locus_listing]": search_query,
+    }
+    r = requests.get(ANTARES_URL, params=params)
+    try:
+        r.raise_for_status()
+    except requests.HTTPError as e:
+        logging.exception(e)
+
+    if r.ok:
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:44]:
```suggestion
    else:
```

File: python/across_api/across/resolve.py [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+        antares_data = json.loads(r.text)
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:45]:
```suggestion
        antares_data = r.json()
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: python/across_api/across/resolve.py [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+            return ra, dec
+    return None, None
+
+
+class Resolve(ACROSSAPIBase):
+    """
+    Resolve class for resolving astronomical object names.
+
+    Parameters:
+    -----------
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:58]:
```suggestion
    Parameters
    ----------
```

File: python/across_api/across/resolve.py [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+    name
+        The name of the astronomical object to resolve.
+
+    Attributes:
+    -----------
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:63]:
```suggestion
    Attributes
    ----------
```

File: python/across_api/across/resolve.py [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+        The right ascension of the resolved object.
+    dec
+        The declination of the resolved object.
+    name
+        The name of the astronomical object.
+    resolver
+        The resolver used for resolving the object.
+
+    Methods:
+    --------
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:74]:
```suggestion
    Methods
    -------
```

File: python/across_api/across/resolve.py [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+    except requests.HTTPError as e:
+        logging.exception(e)
+
+    if r.ok:
+        antares_data = json.loads(r.text)
+        if antares_data["meta"]["count"] > 0:
+            ra = antares_data["data"][0]["attributes"]["ra"]
+            dec = antares_data["data"][0]["attributes"]["dec"]
+            return ra, dec
+    return None, None
```
Comment (lpsinger) [inline python/across_api/across/resolve.py:50]: Ah, we can simplify this a bit. It seems that the ANTARES API will return an HTTP 200 OK status even if no match is found. So we can simply raise for any other status; raising an uncaught exception will result in Lambda returning an HTTP 500 error, which is appropriate.

```suggestion
    r = requests.get(ANTARES_URL, params=params)
    r.raise_for_status()
    antares_data = r.json()
    if antares_data["meta"]["count"] > 0:
        ra = antares_data["data"][0]["attributes"]["ra"]
        dec = antares_data["data"][0]["attributes"]["dec"]
    else:
        ra = dec = None
    return ra, dec
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Review by lpsinger — APPROVED** [review]
Looks great, @jak574.

I don't fully understand the interplay between Pydantic and FastAPI yet, but I'll learn!
